vizro-core/tests/unit/vizro/models/_action/test_action.py

```diff
@@ -282,67 +299,104 @@ def test_action_callback_function_return_value_valid(
-            (["new_value", "new_value_2"], []),
-            (["new_value", "new_value_2"], ["component_1_property", "component_2_property", "component_3_property"]),
-            ({}, []),
-            ({"component_1_property": "new_value"}, []),
-            ({"component_1_property": "new_value"}, ["component_1_property", "component_2_property"]),
-            ({"component_1_property": "new_value", "component_2_property": "new_value_2"}, []),
-            (
-                {"component_1_property": "new_value", "component_2_property": "new_value_2"},
-                ["component_1_property", "component_2_property", "component_3_property"],
-            ),
+            (None, ["component_1_property"]),
+            (False, ["component_1_property"]),
+            (0, ["component_1_property"]),
+            (123, ["component_1_property"]),
         ],
         indirect=["custom_action_function_mock_return"],
     )
-    def test_action_callback_function_return_value_invalid(self, custom_action_function_mock_return, callback_outputs):
+    def test_action_callback_function_outputs_list_return_value_not_collection(
+        self, custom_action_function_mock_return, callback_outputs
+    ):
         action = Action(function=custom_action_function_mock_return())
         with pytest.raises(
             ValueError,
-            match="Number of action's returned elements \\(.?\\)"
-            " does not match the number of action's defined outputs \\(.?\\).",
+            match="Action function has not returned a list or similar but the action's defined outputs are a list.",
         ):
             action._action_callback_function(inputs={}, outputs=callback_outputs)
 
     @pytest.mark.parametrize(
         "custom_action_function_mock_return, callback_outputs",
         [
-            (
-                (namedtuple("Outputs", ["component_1_property"])("new_value")),
-                [],
-            ),
-            (
-                (namedtuple("Outputs", ["component_1_property", "component_2_property"])("new_value", "new_value_2")),
-                ["component_1_property", "component_2_property", "component_3_property"],
-            ),
+            (None, {"component_1_property": "value"}),
+            (False, {"component_1_property": "value"}),
+            (0, {"component_1_property": "value"}),
+            (123, {"component_1_property": "value"}),
+            ("", {"component_1_property": "value"}),
+            ([], {"component_1_property": "value"}),
+            ((), {"component_1_property": "value"}),
+        ],
+        indirect=["custom_action_function_mock_return"],
+    )
+    def test_action_callback_function_outputs_mapping_return_value_not_mapping(
+        self, custom_action_function_mock_return, callback_outputs
+    ):
+        action = Action(function=custom_action_function_mock_return())
+        with pytest.raises(
+            ValueError,
+            match="Action function has not returned a dictionary or similar "
```

Comment:
What do we mean by similar? I assume this is a custom error function right? Should we not be more precise, e.g. say we expect `abc.mapping`?

Reply:
Technically the error message should indeed be "Action function has not return a `Mapping`" but I decided that was too technical/confusing because most users know what a dictionary is but not necessarily a Mapping. So the "or similar" case is a hint towards the fact that really you don't have to return a dictionary, just something that's dictionary-like.

If these seems to add more confusion then I would just remove the "or similar" rather than use the terminology "Mapping" I think.

Reply:
I think dictionary-like might be best! It has the most used word in it, but it also yields mapping to people that know what that is. With "or similar" I think it's a bit confusing whther it could also be say a list or something

Reply:
Good plan! Done in [3052306](https://github.com/mckinsey/vizro/pull/202/commits/30523061ecd43b78d008785ff76967c81800dc49).